### ISSUE TITLE
Service Principal/Enterprise Application support

### ISSUE THREAD

**Issue by idg10** [microsoftgraph/msgraph-sdk-dotnet#675]
I am using the `Microsoft.Graph` library to automate creation and configuration of the AD apps associated with an Azure-based deployment, and one of the things I need to do is assign service principals to application roles. There doesn't appear to be any support for service principals.

In particular, I need to be able to do two things:

* Discover the ID of the service principal corresponding to a particular application in a particular directory (i.e., the object that the Azure Portal shows as the "Enterprise application" - this is the programmatic equivalent of clicking on the **Managed application in local directory** link for an App Registration in the portal)
* assign a service principal to an application role

For the first of these, ideally I'd like to be able to do something like:

```csharp
IGraphServiceServicePrincipalsCollectionPage sp = await client
    .ServicePrincipals
    .Request()
    .Filter($"appId eq '{myAppId}'")
    .GetAsync();
```

But there is no `ServicePrincipals` collection available on `GraphServiceClient`. So instead, I do this:

```csharp
// Service principals don't seem to be supported directly by the SDK yet.
var spReq = new HttpRequestMessage(
    HttpMethod.Get,
    new Uri(new Uri(client.BaseUrl), $"/beta/servicePrincipals?filter=appId eq '{myAppId}'"));
HttpResponseMessage results = await client.HttpProvider.SendAsync(spReq);
results.EnsureSuccessStatusCode();
string response = await results.Content.ReadAsStringAsync();
var jo = JObject.Parse(response);
string spId = jo["value"][0]["id"].ToString();
```

And similarly, I'm having to create my own request to modify the app role assignments. I create a body like this:

```json
{'appRoleId': '<my role id>','principalId': '<service principal id>','resourceId': <enterprise app service principal id>'}
```

and then I POST this to `https://graph.microsoft.com/beta/servicePrincipals/<enterprise app service principal id>/appRoleAssignments`

(In both cases, `<enterprise app service principal id>` is the object id discovered by retrieving the service principal that represents the target application in my tenant.)

The manual coding works, but pretty much everything else is so much easier because of the direct support provided by the Graph SDK. I hope you'll be adding support for service principals at some point soon.

**Comment by kalyankrishna1** [comment]
ServicePrincipals is still in /beta. Should be available in 1.0 (and in Graph SDK) hopefully by June 2020 end.

**Comment by MIchaelMainer** [comment]
Please use the Microsoft.Graph.Beta library to access ServicePrincipals.
